### PR TITLE
Add possibility to perform dimensionality reduction using Umap algorithm

### LINKED ISSUE #109: UMAP
Hi @stefanhahmann et al,

as discussed today, it would be cool to have a way for computing the [UMAP](https://umap-learn.readthedocs.io/en/latest/) of extracted features, to allow dimensionality reduction and viewing complex measurements / data points in 2D. There is a potential Java library for this:
* https://github.com/tag-bio/umap-java

It would be interesting to know if the library can deal with big data. If it does not, I would post-pone developing a big-data version to a later point. For now, it might be most important to have *a* UMAP implementation we can work with to explore possibilities. A potential alternative could also be a python-wrapper using [cuML](https://github.com/rapidsai/cuml), which contains a GPU-accelerated UMAP implementation.

Optionally, we could also think of t-SNE and PCA as alternatives, if not too complicated. 
* https://github.com/lejon/T-SNE-Java
* https://spark.apache.org/docs/2.1.0/mllib-dimensionality-reduction.html#principal-component-analysis-pca

Let me know what you think!

Best,
Robert

### PR DESCRIPTION
**Pull request by stefanhahmann**
(Partially) Resolves #109 

* For the sake of simplicity and also since tests with typical data showed acceptable performance, I decided to not yet go for [cuML](https://github.com/rapidsai/cuml).
* tSNE and PCA are also not yet implemented by this PR. It should be not too complicated to add them with the work done in this PR. However, I would prefer to give this into the hands of users and getting feedback before adding more features/options.

### Description of changes

For getting more insights into users data, this PR adds the possibility to reduce the dimensionality of the measurements, using this algorithm:

* [Uniform Manifold Approximation Projection (UMAP)](https://umap-learn.readthedocs.io/en/latest/)

To apply it to the data, users can call the menu `Plugins > Compute Feature > Dimensionality reduction > UMAP`.
They then need to select the table whose features should be dimensionality reduced. Currently, the Spot and the BranchSpot table are supported. Next, the user needs to select the feature + feature projections that should be dimensionality reduced.

umap_dialog

By default, all measurements are selected in the box.

### Parameters

* Standardize: Whether to standardize the data before reducing the dimensionality. Standardization is recommended when
  the data has different scales / units.
  Further
  reading: [Standardization](https://scikit-learn.org/stable/modules/preprocessing.html#standardization-or-mean-removal-and-variance-scaling).
* Number of dimensions: The number of reduced dimensions to use. The default is 2, but 3 is also common.
  Further reading: [Number of Dimensions](https://umap-learn.readthedocs.io/en/latest/parameters.html#n-components).
* Number of neighbors: The size of the local neighborhood (in terms of number of neighboring sample points) used for
  manifold approximation.
  Larger values result in more global views of the manifold, while smaller values result in more local data being
  preserved.
  In general, it should be in the range 2 to 100.
  Further reading: [Number of Neighbors](https://umap-learn.readthedocs.io/en/latest/parameters.html#n-neighbors).
* Minimum distance: The minimum distance that points are allowed to be apart from each other in the low dimensional
  representation. This parameter controls how tightly UMAP is allowed to pack points together.
  Further reading: [Minimum Distance](https://umap-learn.readthedocs.io/en/latest/parameters.html#min-dist). 

When the user is done with the selection, he/she can click on `Compute UMAP`.
The resulting values will be added as additional columns to the selected table.

umap_table

You can visualize the results using the `Grapher` View of Mastodon and selecting the newly added columns.

umap_grapher

### CODE REVIEW

**Comment by stefanhahmann** [comment]
@haesleinhuepf I added you as a reviewer to this PR, since you drafted the issue that it (partially) resovles: #113

While I am not expecting a code review from your side, a functional review would be interesting. Could also be a joint session, since getting the branch of the plugin to run within Mastodon may take some time.

File: doc/dimensionalityreduction/readme.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+## Dimensionality reduction
+
+For getting more insights into your data, you can reduce the dimensionality of the measurements, using this algorithm:
+
+* [Uniform Manifold Approximation Projection (UMAP)](https://umap-learn.readthedocs.io/en/latest/)
```
Comment (haesleinhuepf) [inline doc/dimensionalityreduction/readme.md:5]:
```suggestion
* [Uniform Manifold Approximation Projection (UMAP)](https://arxiv.org/abs/1802.03426)
* [UMAP Python implementation](https://umap-learn.readthedocs.io/en/latest/)
```

File: doc/dimensionalityreduction/readme.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+## Dimensionality reduction
+
+For getting more insights into your data, you can reduce the dimensionality of the measurements, using this algorithm:
```
Comment (haesleinhuepf) [inline doc/dimensionalityreduction/readme.md:3]:
```suggestion
For visualizing high-dimensional data, e.g. in two dimensions, potentially getting more insights into your data, you can reduce the dimensionality of the measurements, using this algorithm:
```

File: doc/dimensionalityreduction/readme.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+* [Uniform Manifold Approximation Projection (UMAP)](https://umap-learn.readthedocs.io/en/latest/)
+
+To apply it to your data, use the menu `Plugins > Compute Feature > Dimensionality reduction > UMAP`.
+Select the table whose features should be dimensionality reduced. Currently, the Spot and the BranchSpot table are
+supported.
+Next, select the feature + feature projections that should be dimensionality reduced.
```
Comment (haesleinhuepf) [inline doc/dimensionalityreduction/readme.md:10]:
```suggestion
Next, select the feature + feature projections that should be dimensionality reduced. Select only features which describe the phenotype (spot size, shape, speed, number of neighbors, etc.) and avoid selecting technical features (e.g. Spot ID). Only select positional features (centroid) if the position of cells within the image are descriptive for the phenotype.
```

Just some ideas how to guide users. If they cannot select Spot ID anyway, please remove it from the suggestion.

**Review by haesleinhuepf — APPROVED** [review]
Hey @stefanhahmann ,

as discussed today, the functionality looks good to me. I made some minor suggestions about the documentation below. Furthermore, one request: If possible, could we rename "UMAP output dimenstion 0" to "UMAP 0" ? It's pretty standard to call it like this and the shorter, the better.

Otherwise great. Thanks for working on this!

Best,
Robert

File: src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/feature/AbstractUmapFeatureComputer.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	protected void compute( final V vertex )
+	{
+		if ( isVertexValid( vertex ) )
+		{
+			for ( int i = 0; i < settings.getNumberOfOutputDimensions(); i++ )
+			{
+				DoublePropertyMap< V > umapOutput = feature.getUmapOutputMaps().get( i );
+				umapOutput.set( vertex, umapResult[ index ][ i ] );
+			}
+			index++;
```
Comment (maarzt) [inline src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/feature/AbstractUmapFeatureComputer.java:115]: The usage of this "index" field in the class is horribly error prone. PLEASE replace it with a vertexToRowId map!!!! (You can use `RefToIntMap` or something like that.)

File: src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/feature/AbstractUmapFeatureComputer.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+		logger.info( "Fitting umap. Data matrix has {} rows x {} columns.", dataMatrix.length, inputDimensions.size() );
+		umapResult = umap.fitTransform( dataMatrix );
+		logger.info( "Finished fitting umap. Results has {} rows x {} columns.", umapResult.length,
+				umapResult.length > 0 ? umapResult[ 0 ].length : 0 );
+	}
+
+	private List< double[] > extractAndCacheValidDataRows()
+	{
+		List< double[] > data = new ArrayList<>();
+		for ( V vertex : getVertices() )
```
Comment (maarzt) [inline src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/feature/AbstractUmapFeatureComputer.java:176]: You are using a `Collection<V>` interface here. Don't assume that there is a particular iterations order for a `Collection` it's not guarantied.

**Review by maarzt — CHANGES_REQUESTED** [review]
There are big problems that still need to be fixed.

File: src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/util/StandardScalerTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+				{ 3, 3, 3 },
+				{ 3, 3, 3 }
+		};
+		double[][] expected = {
+				{ 0, 0, 0 },
+				{ 0, 0, 0 },
+				{ 0, 0, 0 }
+		};
+		StandardScaler.standardizeColumns( array );
+		assertArrayEquals( expected, array );
```
Comment (maarzt) [inline src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/util/StandardScalerTest.java:83]: I would prefer if only one of the three columns has variance zero. To make sure the other columns are still treated correctly.

File: src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/util/StandardScaler.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+		for ( int i = 0; i < numRows; i++ )
+		{
+			column[ i ] = matrix[ i ][ columnIndex ];
+		}
+		double[] standardizedColumn = StatUtils.normalize( column );
+
+		for ( int i = 0; i < numRows; i++ )
+		{
+			matrix[ i ][ columnIndex ] = Double.isNaN( standardizedColumn[ i ] ) ? 0d : standardizedColumn[ i ]; // replace NaNs with 0
```
Comment (maarzt) [inline src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/util/StandardScaler.java:41]: This line removes any NaN value in the matrix. I guess it is implemented like this on the assumption, that all NaN values originate from StatUtils.normalize when applied on a constant column. It is ok to make this assumption, but only if you assume that the input matrix has NaN. Which is also true in this code base. But please document, that the standardizeColumns method expects the input matrix/array to have no NaN values.

Comment (maarzt) [reply src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/util/StandardScaler.java:41]: This line could benefit from a comment, like:
_StatUtils.normalize(...) will write NaN if a column has variance=0, it is important to replace this NaN values before computing the UMAP_

**Review by maarzt — APPROVED** [review]
Some small remarks.

**Pull request merged**